posthog/caching/insight_caching_state.py

```diff
@@ -0,0 +1,136 @@
+            last_viewed_at__gte=now() - VERY_RECENTLY_VIEWED_THRESHOLD
+        ).distinct("insight_id")
+        return set(recently_viewed_insights.values_list("insight_id", flat=True))
+
+
+def upsert(
+    team: Team, target: Union[DashboardTile, Insight], lazy_loader: Optional[LazyLoader] = None
+) -> Optional[InsightCachingState]:
+    lazy_loader = lazy_loader or LazyLoader()
+    cache_key = calculate_cache_key(team, target)
+    if cache_key is None:  # Non-cachable model
+        return None
+
+    target_age = calculate_target_age(team, target, lazy_loader)
+    target_age_seconds = target_age.value.total_seconds() if target_age.value is not None else None
+    caching_state, _ = InsightCachingState.objects.update_or_create(
+        team_id=team.pk,
+        insight=target if isinstance(target, Insight) else target.insight,
+        dashboard_tile=target if isinstance(target, DashboardTile) else None,
+        defaults={
+            "cache_key": cache_key,
+            "target_cache_age_seconds": target_age_seconds,
+        },
+    )
+    return caching_state
+
+
+def fetch_states_in_need_of_updating():
+    return InsightCachingState.objects.raw(
+        """
+        SELECT *
+        FROM posthog_insightcachingstate
+        WHERE target_cache_age_seconds IS NOT NULL
+          AND refresh_attempt < %(max_attempts)s
+          AND (
+            last_refresh IS NULL OR
+            last_refresh < %(timestamp)s - target_cache_age_seconds * interval '1' second
+          )
+        """,
+        {"max_attempts": MAX_ATTEMPTS, "timestamp": now()},
+    )
+
+
+def calculate_cache_key(team: Team, target: Union[DashboardTile, Insight]) -> Optional[str]:
+    insight = target if isinstance(target, Insight) else target.insight
+    if insight is None:
+        return None
+
+    return generate_insight_cache_key(insight, insight.dashboard)
+
+
+def calculate_target_age(team: Team, target: Union[DashboardTile, Insight], lazy_loader: LazyLoader) -> TargetCacheAge:
+    if isinstance(target, Insight):
+        return calculate_target_age_insight(team, target, lazy_loader)
+    else:
+        return calculate_target_age_dashboard_tile(team, target, lazy_loader)
+
+
+def calculate_target_age_insight(team: Team, insight: Insight, lazy_loader: LazyLoader) -> TargetCacheAge:
+    if team.pk not in lazy_loader.active_teams:
+        return TargetCacheAge.NO_CACHING
+
+    if insight.deleted or len(insight.filters) == 0:
+        return TargetCacheAge.NO_CACHING
+
+    # :TODO: Only cache insights that are shared
+    if insight.pk not in lazy_loader.recently_viewed_insights:
+        return TargetCacheAge.NO_CACHING
+
+    return TargetCacheAge.MID_PRIORITY
+
+
+def calculate_target_age_dashboard_tile(
+    team: Team, dashboard_tile: DashboardTile, lazy_loader: LazyLoader
+) -> TargetCacheAge:
+    if team.pk not in lazy_loader.active_teams:
+        return TargetCacheAge.NO_CACHING
+
+    if dashboard_tile.dashboard.deleted:
```

Comment:
the tile can be deleted too

TBH soft deletion of the insight or dashboard should be triggering soft-deletion of the tile, so we're (theoretically) safe to only check if the tile is deleted.

Reply:
We shouldn't be baking that assumption in here - if there isn't a DB constraint or something to enforce it, needing to be aware of UI contracts here will just lead to bugs down the line.